### ISSUE TITLE
Causal AutoRegressive Doubt

### ISSUE THREAD

**Issue by HaldiramSharma** [lucidrains/performer-pytorch#21]
```
def causal_linear_attention_noncuda(q, k, v):
    k_cumsum = k.cumsum(dim=-2)
    context = torch.einsum('...nd,...ne->...nde', k, v)
    context = context.cumsum(dim=-3)
    context /= k_cumsum.unsqueeze(dim=-1)
    out = torch.einsum('...nde,...nd->...ne', context, q)
    return out
```

I'm not able to understand why context is being divided by k_cumsum?  I understand the one done in bidirectional as  it is mentioned in the paper.
```
D_inv = 1. / torch.einsum('...nd,...d->...n', q, k.sum(dim = -2))
```
but it didn't mention anything about the normalization of the causal attention

Can you please help me understand it? Any help would be highly appreciated. :)

**Comment by lucidrains** [comment]
@HaldiramSharma I think the Performer paper is a bit hard to read. A better paper to read would be the Transformers are RNNs paper https://arxiv.org/abs/2006.16236 You can see how it is derived in equations 10 and 11

As for the performer paper, you can see in section A.3 (and A.4) that renormalization of attention is always turned on

**Comment by HaldiramSharma** [comment]
So i had a look at the Transformers are RNNs paper and according to my finding you might have missed it by mul it by q_prime
```
1/torch.einsum('bhqf, bhkf -> bhqf', q,  k.cumsum(dim=-2))
```

instead of just normalizing by

```
k_cumsum = k.cumsum(dim=-2)
````

not sure if its correct though

**Comment by lucidrains** [comment]
@HaldiramSharma I reread the equations and I believe you are correct! Fixed in 0.7.5! https://github.com/lucidrains/performer-pytorch/releases/tag/0.7.5
